python/felis/metadata.py

```diff
@@ -0,0 +1,500 @@
+            check_obj: dm.CheckConstraint = constraint_obj
+            expression = check_obj.expression
+            constraint = CheckConstraint(expression, **args)
+        elif constraint_type == "Unique" and isinstance(constraint_obj, dm.UniqueConstraint):
+            uniq_obj: dm.UniqueConstraint = constraint_obj
+            columns = [self._objects[column_id] for column_id in uniq_obj.columns]
+            constraint = UniqueConstraint(*columns, **args)
+        else:
+            raise ValueError(f"Unexpected constraint type: {constraint_type}")
+
+        self._objects[constraint_obj.id] = constraint
+
+        return constraint
+
+    def build_index(self, index_obj: dm.Index) -> Index:
+        """Build a SQA `Index` from a `felis.datamodel.Index` object.
+
+        Parameters
+        ----------
+        index_obj : `felis.datamodel.Index`
+            The index object from which to build the SQA index.
+
+        Returns
+        -------
+        index: `Index`
+            The SQA index object.
+        """
+        columns = [self._objects[c_id] for c_id in (index_obj.columns if index_obj.columns else [])]
+        expressions = index_obj.expressions if index_obj.expressions else []
+        index = Index(index_obj.name, *columns, *expressions)
+        self._objects[index_obj.id] = index
+        return index
+
+
+class ConnectionWrapper:
+    """A wrapper for a SQLAlchemy engine or mock connection which provides a
+    consistent interface for executing SQL statements.
+    """
+
+    def __init__(self, engine: Engine | MockConnection):
+        """Initialize the connection wrapper.
+
+        Parameters
+        ----------
+        engine : `Engine` or `MockConnection`
+            The SQLAlchemy engine or mock connection to wrap.
+        """
+        self.engine = engine
+
+    def execute(self, statement: Any) -> ResultProxy:
+        """Execute a SQL statement on the engine and return the result."""
+        if isinstance(statement, str):
+            statement = text(statement)
+        if isinstance(self.engine, MockConnection):
+            return self.engine.connect().execute(statement)
+        else:
+            with self.engine.connect() as connection:
+                result = connection.execute(statement)
+                connection.commit()
+                return result
+
+
+class DatabaseContext:
+    """A class for managing the schema and its database connection."""
+
+    def __init__(self, metadata: MetaData, engine: Engine | MockConnection):
+        """Initialize the database context.
+
+        Parameters
+        ----------
+        metadata : `MetaData`
+            The SQLAlchemy metadata object.
+
+        engine : `Engine` or `MockConnection`
+            The SQLAlchemy engine or mock connection object.
+        """
+        self.engine = engine
+        self.metadata = metadata
+        self.connection = ConnectionWrapper(engine)
+
+    def create_if_not_exists(self) -> None:
```

Comment:
Is there a function on SQA MetaData that does this?

Reply:
> A question for @andy-slac I think.

I'll ask - I couldn't find anything in the SQLAlchemy documentation for `MetaData` which does this.

Reply:
MetaData is for table schema only, to create databases/schemas you need to write the code yourself, and it's not super-portable, so the code in this method looks OK.